src/combinator.rs

```diff
@@ -1618,6 +1615,61 @@ where
     go_extra!(C);
 }
 
+/// See [`IterParser::collect_exactly`]
+pub struct CollectExactly<A, O, C> {
+    pub(crate) parser: A,
+    pub(crate) phantom: PhantomData<(O, C)>,
+}
+
+impl<A: Copy, O, C> Copy for CollectExactly<A, O, C> {}
+impl<A: Clone, O, C> Clone for CollectExactly<A, O, C> {
+    fn clone(&self) -> Self {
+        Self {
+            parser: self.parser.clone(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a, I, O, E, A, C> Parser<'a, I, C, E> for CollectExactly<A, O, C>
+where
+    I: Input<'a>,
+    E: ParserExtra<'a, I>,
+    A: IterParser<'a, I, O, E>,
+    C: ContainerExactly<O>,
+{
+    #[inline]
+    fn go<M: Mode>(&self, inp: &mut InputRef<'a, '_, I, E>) -> PResult<M, C> {
+        let before = inp.offset();
+        let mut output = M::bind(|| C::uninit());
+        let mut iter_state = self.parser.make_iter::<M>(inp)?;
+        for idx in 0..C::LEN {
+            match self.parser.next::<M>(inp, &mut iter_state) {
+                Ok(Some(out)) => {
+                    M::combine_mut(&mut output, out, |c, out| C::write(c, idx, out));
+                }
+                Ok(None) => {
+                    inp.add_alt(
+                        inp.offset,
+                        None,
+                        None,
+                        unsafe { inp.span_since(before) },
+                    );
+                    M::map(output, |mut output| unsafe { C::drop_before(&mut output, idx) });
+                    return Err(())
+                }
+                Err(()) => {
+                    M::map(output, |mut output| unsafe { C::drop_before(&mut output, idx) });
```

Comment:
Maybe a `map_mut` that takes `&mut output` (like `combine_mut`) would speed up this case?